### PR TITLE
General improvement opportunities

### PR DESCRIPTION
**Pull request by DanielAtKrypton**
Changed file name to csv2npz.

### CODE REVIEW

**Comment by DanielAtKrypton** [comment]
The main improvement here is the automatic download of challenge files. See [README](https://github.com/DanielAtKrypton/ozechallenge_benchmark/blob/patch-1/README.md) for instructions on how to configure credentials.

There are other improvements though.
For example:

- Improved `README.md`.
- Hard coded values inside `benchmark.ipynb` are now variables.
- Implementation of `get_x_shape` and `get_y_shape` methods.
- Add option to use the transformer in `benchmark.ipynb`.
- Code linted.

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Environment
 .env
+.env.test.local
```
Comment (maxjcohen) [inline .gitignore:3]: What is this line ignoring ?

File: requirements.txt [hunk 2]
```diff
@@ -1,64 +1,12 @@
...
+matplotlib
+seaborn
+tqdm
+pylint
+autopep8
+pytest
+python-dotenv
+python-dotenv[cli]
+psutil
+lxml
```
Comment (maxjcohen) [inline requirements.txt]: Yes this definitely should had been done already.

File: src/dataset.py [hunk 3]
```diff
@@ -35,7 +35,7 @@ class OzeEvaluationDataset(Dataset):
         Normalization constant.
     """
 
-    def __init__(self, dataset_x_path, labels_path="labels.json", **kwargs):
+    def __init__(self, dataset_x_path, time_series_length, labels_path="labels.json", **kwargs):
```
Comment (maxjcohen) [inline src/dataset.py]: Can you add a default value of 672, as well a description in the Parameters section ?

File: src/model.py [hunk 4]
```diff
@@ -41,6 +40,7 @@ def __init__(self, input_dim=37, hidden_dim=100, output_dim=8, num_layers=3, **k
         self.lstm = nn.LSTM(input_dim, hidden_dim, num_layers=num_layers, batch_first=True)
         self.linear = nn.Linear(hidden_dim, output_dim)
 
+    # NOTE Parameters differ from overridden 'forward' methodpylint(arguments-differ)
```
Comment (maxjcohen) [inline src/model.py]: As for the other repo, let's keep pylint messages outside of the python files, especially for this code which is more subjected to being read than executed.

File: time_series_transformer/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+"""
+__init__.py
+"""
+from .transformer import Transformer
```
Comment (maxjcohen) [inline time_series_transformer/__init__.py]: We do not wish to create a bias for the challenge by adding the transformer code, this is why we kept those two repos independent, so the time_series_transformer folder should be remove.

If you want to use the transformer code in this repo, you could add `git+https://github.com/maxjcohen/transformer` to your `requirements.txt`.

File: src/dataset.py [hunk 6]
```diff
@@ -101,7 +103,7 @@ class OzeDataset(OzeEvaluationDataset):
         Dataset target of shape (m, K, 8).
     """
 
-    def __init__(self, dataset_x_path, dataset_y_path, labels_path="labels.json", **kwargs):
+    def __init__(self, dataset_x_path, dataset_y_path, time_series_length, labels_path="labels.json", **kwargs):
```
Comment (maxjcohen) [inline src/dataset.py]: Same remark for the default value

File: src/utils.py [hunk 7]
```diff
@@ -15,21 +200,21 @@ def csv2npz(dataset_x_path, dataset_y_path, labels_path='labels.json'):
...
-            for i in range(K)]]
+           for i in range(K)]]
     Z = Z.values.reshape((m, -1, K))
 #     Z = Z.transpose((0, 2, 1))
     Z = Z.astype(np.float32)
-    
-    np.savez('dataset.npz', R=R, X=X, Z=Z)
\ No newline at end of file
+
+    np.savez(path.join(output_path, filename), R=R, X=X, Z=Z)
```
Comment (maxjcohen) [inline src/utils.py:232]: The path joining could be done with the pathlib instead, for instance
```python
Path(output_path).joinpath(filename)
```

**Review by maxjcohen — CHANGES_REQUESTED** [review]
Hi, again a good initiative, there are several issues in this repo that need solving. I have added a few comments where we need changes before merging, the rest in good.

The main points to attend to:
- Removing the transformer folder (and the modifications to the benchmark notebook regarding transformer use). For your personal use, you can add the other repo as dependency, see the comment.
- Adding parameters instead of hard coded values is an improvement, but we should add default values.

File: src/utils.py (hunk 7 above)
Comment (DanielAtKrypton) [reply src/utils.py:232]: Makes sense.

File: time_series_transformer/__init__.py (hunk 5 above)
Comment (DanielAtKrypton) [reply time_series_transformer/__init__.py]: Makes sense.

File: src/model.py (hunk 4 above)
Comment (DanielAtKrypton) [reply src/model.py]: Makes sense.

File: .gitignore (hunk 1 above)
Comment (DanielAtKrypton) [reply .gitignore:3]: .env.test.local is the file in which user's private credentials to login at http://challengedata.ens.fr/ are stored. This is necessary to enable the subsequent download of the dataset if none is found by npz_check function. More info [here](https://github.com/DanielAtKrypton/ozechallenge_benchmark#dot-env-environment-variables). So this file should be only present locally.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+### Installation
+
+Clone repository with the following command in the terminal
+
+```terminal
+git clone https://github.com/maxjcohen/ozechallenge_benchmark.git
+cd ozechallenge_benchmark
+virtualenv -p python3 .env
+pip install -r requirements
```
Comment (maxjcohen) [inline README.md:115]: Here we should add a line to activate the environment, I had issue with some developers not being used to activate it once created.
```bash
virtualenv -p python3 .env
source .env/bin/activate
pip install -r requirements
```

File: src/dataset.py (hunk 6 above)
Comment (DanielAtKrypton) [reply src/dataset.py]: Alright

File: README.md (hunk 8 above)
Comment (DanielAtKrypton) [reply README.md:115]: Alright.

File: src/dataset.py (hunk 3 above)
Comment (DanielAtKrypton) [reply src/dataset.py]: Sure.

**Comment by DanielAtKrypton** [comment]
> there are several issues in this repo that need solving

I tried to address opportunities from my point of view. Adding GitHub issues might be worth the time in order to allow developers to better understand current concerns.

The main improvement as of now I could think of was to add pytests.
Next step would be to add continuous integration so whenever a commit is made, it automatically triggers pytest and the training would take place within the server.

I believe the evolution of this repository can help the evolution of transformer repository too.
In my opinion, the creation of the [TimeSeriesPredictor](https://github.com/DanielAtKrypton/ozechallenge_benchmark/blob/patch-1/src/time_series_predictor.py#L17) and [LSTMTimeSeriesPredictor](https://github.com/DanielAtKrypton/ozechallenge_benchmark/blob/master/src/transformer_tsp.py#L10) classes isolated concerns nicely.

After your guidance, maybe it is time to get back to the transformer repository and add pytests and continuous integration. I was reflecting on the later and I think there we could simply use a seaborn dataset for py testing. Once that is alright we can remove all transformer related stuff from this repository as per your request. [TransformerTimeSeriesPredictor](https://github.com/DanielAtKrypton/ozechallenge_benchmark/blob/master/src/transformer_tsp.pyL10) class can be useful there.

It may be also worth to create a python package for the `TimeSeriesPredictor`.

**Comment by DanielAtKrypton** [comment]
I notice how faster the `LSTMTimeSeriesPredictor` performs when compared to the `TransformerTimeSeriesPredictor`. I think that might be a small detail related to GPU processing in the [transformer repository](https://github.com/maxjcohen/transformer).

The following image gives and overview of the `TransformerTimeSeriesPredictor` resources consumption even when using `cuda:0`:

File: requirements.txt [hunk 9]
```diff
@@ -1,64 +1,13 @@
...
+matplotlib
+seaborn
+tqdm
+pylint
+autopep8
+python-dotenv
+python-dotenv[cli]
+psutil
+lxml
+git+https://github.com/maxjcohen/transformer
```
Comment (maxjcohen) [inline requirements.txt]: When I said you could add this line locally, i meant it shouldn't be added in the repo's requirements. Indeed, running the benchmark does not require to install or use transformers.

File: src/lstm_tsp.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
+"""
+lstm_tsp
+"""
+import torch
+
+from .model import BenchmarkLSTM
+from .time_series_predictor import TimeSeriesPredictor
+
+class LSTMTimeSeriesPredictor(TimeSeriesPredictor):
```
Comment (maxjcohen) [inline src/lstm_tsp.py]: I don't understand the need for this class

File: src/model.py [hunk 11]
```diff
@@ -41,6 +40,7 @@ def __init__(self, input_dim=37, hidden_dim=100, output_dim=8, num_layers=3, **k
         self.lstm = nn.LSTM(input_dim, hidden_dim, num_layers=num_layers, batch_first=True)
         self.linear = nn.Linear(hidden_dim, output_dim)
 
+    # pylint: disable=arguments-differ
```
Comment (maxjcohen) [inline src/model.py]: Pylint messages should be kept outside the source code, to improve readability

File: src/time_series_predictor.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+import psutil
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+# Show switch to cpu warning
+warnings.filterwarnings("default", category=ResourceWarning)
+
+# pylint: disable=too-many-instance-attributes
+class TimeSeriesPredictor:
```
Comment (maxjcohen) [inline src/time_series_predictor.py]: For a real development project, you could argue this class makes the code cleaner, but in our case it just adds more code without real change in the results.

File: src/transformer_tsp.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
+"""
+transformer_tsp
+"""
+import torch
+from tst import Transformer
+
+from .time_series_predictor import TimeSeriesPredictor
+
+# pylint: disable=too-many-instance-attributes
+class TransformerTimeSeriesPredictor(TimeSeriesPredictor):
```
Comment (maxjcohen) [inline src/transformer_tsp.py]: As said before, I'd like to avoid transformer being mentioned in this repo

**Review by maxjcohen — COMMENTED** [review]
Hi, sorry for the delay in answering your updates. There are still a few modifications needed before merging, mainly:

- removing LSTMTimeSeriesPredictor and likes : these classes neither improve readability or results. They focus on improving the code itself, which is beyond the scope of this repo, as we aim here at demonstrating a benchmark solution to the challenge. Don't forget the target audience for this challenge are students in statistics that may not have extensive development knowledge.
- removing test : same as for the previous point, they are beyond the scope of this repo. We aim at keeping the code as simple to understand for non programmers as possible.
- removing transformer ref : these two repos should really stay independant. Anyone interested in using transformer for the challenge can navigate through our github account and easily make the connection. This is especially true now that you've added instructions as to convert `csv` to `npz`, etc.

Concerning implementation of tests, I'll consider adding them to the transformer repo, where they would be a nice addition. As for issues, they seem to be activated on both repos.

Concerning LSTM being faster than transformer, I have witnessed the same results. This could be due to a number of factors, for instance:

- LSTM having a linear complexity in the time dimension, while the transformer in quadratic
- LSTM are coded in the pytorch library, and thus optimized to run on the GPU, which is not the case for the transformer written in python. If you want to speed up the process, you may want to use pytorch's [Multi Head Attention block](https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html).

File: src/transformer_tsp.py (hunk 13 above)
Comment (DanielAtKrypton) [reply src/transformer_tsp.py]: Got it!

File: src/time_series_predictor.py (hunk 12 above)
Comment (DanielAtKrypton) [reply src/time_series_predictor.py]: Indeed.

File: src/model.py (hunk 11 above)
Comment (DanielAtKrypton) [reply src/model.py]: The purpose here was workaround the arguments differ warning from pylint. I am removing for the sake of simplicity as per your vision.

File: src/lstm_tsp.py (hunk 10 above)
Comment (DanielAtKrypton) [reply src/lstm_tsp.py]: The purpose was to isolate concerns and make a time series predictor resembling [sklearn fit and predict architecture](https://scikit-learn.org/stable/tutorial/statistical_inference/supervised_learning.html). Pros and cons here.

File: requirements.txt (hunk 9 above)
Comment (DanielAtKrypton) [reply requirements.txt]: Got it!

**Comment by DanielAtKrypton** [comment]
> Hi, sorry for the delay in answering your updates.

I completely understand and respect your timings. I am very happy to be able to contribute and learn with such state-of-the-art technology when it comes to time series forecast. Kudos to you for that!

File: src/loss.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
+"""
+Loss
+"""
+import torch
+import torch.nn as nn
+
+
+class OZELoss(nn.Module):
```
Comment (maxjcohen) [inline src/loss.py]: This loss is misleading, as it is not the one used for the challenge (which is the usual MSE). I defined this new loss later based on the literature in building management optimization.

For this repo, I think it's better we remove it. As for the transformer repo, we should rename it and improve the documentation.

Comment (DanielAtKrypton) [reply src/loss.py]: Makes sense.

**Comment by maxjcohen** [comment]
Okay, this PL seems good now, so without further ado, i'll merge it. Thanks again for your contribution !

**Review by maxjcohen — APPROVED** [review]
